File: python/307.py
class Node:
    __slot__ = "l r lt rt s"

    def __init__(self, l, r):
        self.l = l
        self.r = r


class NumArray(object):

    @staticmethod
    def _build(l, r, nums):
        root = Node(l, r)
        if l == r:
            root.s = nums[l]
        else:
            m = (l+r) // 2
            root.lt = NumArray._build(l, m, nums)
            root.rt = NumArray._build(m+1, r, nums)
            root.s = root.lt.s + root.rt.s
        return root

    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        if not nums:
            return
        self.root = NumArray._build(0, len(nums)-1, nums)

    def update(self, i, val, root=None):
        """
        :type i: int
        :type val: int
        :rtype: int
        """
        if root is None:
            root = self.root
        if root.l == root.r:
            root.s = val
            return
        m = (root.l + root.r) // 2
        if i <= m:
            self.update(i, val, root.lt)
        else:
            self.update(i, val, root.rt)
        root.s = root.lt.s + root.rt.s

    def sumRange(self, i, j, root=None):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        if root is None:
            root = self.root
        if i == root.l and j == root.r:
            return root.s
        m = (root.l + root.r) // 2
        if j <= m:
            return self.sumRange(i, j, root.lt)
        elif i > m:
            return self.sumRange(i, j, root.rt)
        else:
            return self.sumRange(i, m, root.lt) +\
                   self.sumRange(m+1, j, root.rt)

File: cpp/307.cpp
#include <vector>
#include <iostream>
using namespace std;

struct Node {int l, r, s;};

class NumArray {
  Node tree[100000];

  void build(int l, int r, vector<int> &nums, int rt){
    tree[rt].l = l;
    tree[rt].r = r;
    if(l == r)
      tree[rt].s = nums[l];
    else{
      int m = (l + r) / 2;
      build(l, m, nums, 2*rt);
      build(m+1, r, nums, 2*rt+1);
      tree[rt].s = tree[2*rt].s + tree[2*rt+1].s;
    }
  }

public:
  NumArray(vector<int> &nums) {
    if(nums.empty()) return;
    build(0, nums.size()-1, nums, 1);
  }

  void update(int i, int val, int rt=1) {
    if(tree[rt].l == tree[rt].r){
      tree[rt].s = val;
      return;
    }
    int m = (tree[rt].l + tree[rt].r) / 2;
    if(i <= m)
      update(i, val, rt*2);
    else
      update(i, val, rt*2+1);
    tree[rt].s = tree[rt*2].s + tree[rt*2+1].s;
  }

  int sumRange(int i, int j, int rt=1) {
    if (i == tree[rt].l && j == tree[rt].r)
      return tree[rt].s;
    int m = (tree[rt].l + tree[rt].r) / 2;
    if(j <= m)
      return sumRange(i, j, rt*2);
    else if(i > m)
      return sumRange(i, j, rt*2+1);
    else return sumRange(i, m, rt*2) + sumRange(m+1, j, rt*2+1);
  }
};

int main(){
  vector<int> v {1, 3, 5};
  NumArray sa(v);
  cout << sa.sumRange(0, 2) << endl;
}
